### PR TITLE
feat(bedrock): support metrics for bedrock

### PR DESCRIPTION
**Pull request by jinsongo**
- [x] I have added tests that cover my changes.
- [ ] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [x] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.

### CODE REVIEW

**Comment by jinsongo** [comment]
@nirga Could you please advise why I always get the following error, even though I tried changing the model ID? Thanks.

https://github.com/traceloop/openllmetry/actions/runs/10912918948/job/30288462269?pr=1957
`botocore.errorfactory.AccessDeniedException: An error occurred (AccessDeniedException) when calling the InvokeModel operation: You don't have access to the model with the specified model ID.`    cc @gyliu513

**Comment by jinsongo** [comment]
I already know the root cause why I always getting the error `botocore.errorfactory.AccessDeniedException: An error occurred (AccessDeniedException) when calling the InvokeModel operation: You don't have access to the model with the specified model ID.`. Thanks.

File: packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py [hunk 1]
```diff
@@ -29,6 +29,8 @@ class Meters:
     LLM_WATSONX_COMPLETIONS_RESPONSES = "llm.watsonx.completions.responses"
     LLM_WATSONX_COMPLETIONS_TOKENS = "llm.watsonx.completions.tokens"
 
+    LLM_BEDROCK_COMPLETIONS_EXCEPTIONS = "llm.bedrock.completions.exceptions"
```
Comment (gyliu513) [inline packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:32]: Is this required?

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py [hunk 2]
```diff
@@ -62,36 +89,72 @@ def _set_span_attribute(span, name, value):
 def _with_tracer_wrapper(func):
     """Helper for providing tracer for wrapper functions."""
 
-    def _with_tracer(tracer, to_wrap):
+    def _with_tracer(
+        tracer,
+        metricParams,
```
Comment (gyliu513) [inline packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:94]: If you are adding a new parameter, I prefer put it as last one.

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py [hunk 3]
```diff
@@ -62,36 +89,72 @@ def _set_span_attribute(span, name, value):
...
         return wrapper
 
     return _with_tracer
 
 
 @_with_tracer_wrapper
-def _wrap(tracer, to_wrap, wrapped, instance, args, kwargs):
+def _wrap(
+    tracer,
+    metricParams,
```
Comment (gyliu513) [inline packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:116]: ditto

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py [hunk 4]
```diff
@@ -476,6 +589,34 @@ def _set_amazon_span_attributes(span, request_body, response_body):
             )
 
 
+def _create_metrics(meter: Meter):
```
Comment (gyliu513) [inline packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:592]: There are duplicate code in anthropic and groq, we may need a clean up in future, like creating a common shared lib for the LLM instrumentation. @nirga ^^

File: packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py (hunk 1 above)
Comment (jinsongo) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:32]: It's an open issue. Actually, I think it's better to use a common name, for example `llm.completions.exceptions` for all ai system to count exceptions. If yes, I would like to use another PR to fix all related code.

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py (hunk 2 above)
Comment (jinsongo) [reply packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:94]: I just want to put the metric parameters near by tracer and it's an internal function only.

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py (hunk 4 above)
Comment (nirga) [reply packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:592]: Yep! I agree
https://github.com/traceloop/openllmetry/issues/1997

File: packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py (hunk 1 above)
Comment (nirga) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:32]: @jinsongo if we're doing this here I'd try to use the new genAI conventions - https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics
I think they're already released so you can even use them directly from the original otel semconv package

Comment (nirga) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:32]: (in this case - it means exceptions are counted in the duration metric)

Comment (jinsongo) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:32]: @nirga I cannot find an appropriate metric name for the exceptions counter of Bedrock. As you know, version 0.4.1 of opentelemetry-semantic-conventions-ai does not include `LLM_BEDROCK_COMPLETIONS_EXCEPTIONS` although it's already included in this current PR commits, which causes the build to fail when I use it in the Bedrock instrumentation code. As you know, that's the same approach for OpenAI, Authropic and Watsonx.

Comment (nirga) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:32]: @jinsongo this should be in the standard `opentelemetry-semantic-conventions` package. And you should count exceptions on the duration metrics as specified in the spec.

Comment (gyliu513) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py:32]: @jinsongo If this parameter is not required, I think we can ignore this for now, and we can fix this in future as a consolidation for semantic convention.

Comment (jinsongo) [reply packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv_ai/__init__.py]: @gyliu513 I have removed `LLM_BEDROCK_COMPLETIONS_EXCEPTIONS` from `semconv_ai`, and added a `TODO` comment about fixing in future as a consolidation for semantic convention.

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py (hunk 4 above)
Comment (jinsongo) [reply packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:592]: Will fix with another PR once the code for https://github.com/traceloop/openllmetry/issues/1997 is ready.

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py [hunk 5]
```diff
@@ -27,9 +29,30 @@
...
+    def __init__(
+        self,
+        token_histogram: Histogram,
+        choice_counter: Counter,
+        duration_histogram: Histogram,
+        exception_counter: Counter,
+    ):
+        self.vendor = ""
+        self.model = ""
+        self.isStream = False
```
Comment (nirga) [inline packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:48]: nit: `is_stream`

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py (hunk 2 above)
Comment (nirga) [inline packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py:94]: nit: `metric_params`

**Review by nirga — COMMENTED** [review]
nice @jinsongo! a couple of styling comments and we're ready to go

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py (hunk 2 above)
Comment (jinsongo) [reply packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py]: Done

File: packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py (hunk 5 above)
Comment (jinsongo) [reply packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py]: Done

**Comment by gyliu513** [comment]
/lgtm

@nirga can you help review? Thanks

**Comment by jinsongo** [comment]
@nirga Do you have more comments? Thanks.

**Review by nirga — APPROVED** [review]

**Pull request merged**